pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -0,0 +1,2272 @@
+                                                                                discrete_only=False)
+            elif config.add_regularization in {'grad_lag', 'hess_lag', 'hess_only_lag', 'sqp_lag'}:
+                MindtPy.roa_proj_mip_obj = generate_lag_objective_function(self.mip,
+                                                                        self.best_solution_found,
+                                                                        config,
+                                                                        self.timing,
+                                                                        discrete_only=False)
+            if self.objective_sense == minimize:
+                MindtPy.cuts.obj_reg_estimate = Constraint(
+                    expr=sum(MindtPy.objective_value[:]) <= (1 - config.level_coef) * self.primal_bound + config.level_coef * self.dual_bound)
+            else:
+                MindtPy.cuts.obj_reg_estimate = Constraint(
+                    expr=sum(MindtPy.objective_value[:]) >= (1 - config.level_coef) * self.primal_bound + config.level_coef * self.dual_bound)
+        else:
+            if config.add_slack:
+                MindtPy.del_component('aug_penalty_expr')
+
+                MindtPy.aug_penalty_expr = Expression(
+                    expr=sign_adjust * config.OA_penalty_factor * sum(
+                        v for v in MindtPy.cuts.slack_vars[...]))
+            main_objective = MindtPy.objective_list[-1]
+            MindtPy.mip_obj = Objective(
+                expr=main_objective.expr +
+                (MindtPy.aug_penalty_expr if config.add_slack else 0),
+                sense=self.objective_sense)
+
+            if config.use_dual_bound:
+                # Delete previously added dual bound constraint
+                MindtPy.cuts.del_component('dual_bound')
+                if self.dual_bound not in {float('inf'), float('-inf')}:
+                    if self.objective_sense == minimize:
+                        MindtPy.cuts.dual_bound = Constraint(
+                            expr=main_objective.expr +
+                            (MindtPy.aug_penalty_expr if config.add_slack else 0) >= self.dual_bound,
+                            doc='Objective function expression should improve on the best found dual bound')
+                    else:
+                        MindtPy.cuts.dual_bound = Constraint(
+                            expr=main_objective.expr +
+                            (MindtPy.aug_penalty_expr if config.add_slack else 0) <= self.dual_bound,
+                            doc='Objective function expression should improve on the best found dual bound')
+
+
+    def export_attributes(self):
+        for name, val in self.__dict__.items():
+            setattr(self.solve_data, name, val) 
+
+    def update_attributes(self):
+        for name, val in self.solve_data.__dict__.items():
+            self.__dict__[name] = val
```

Comment:
This is because we can not access the attributes in single tree implementation. Therefore, we used the `solve_data` way and keep `solve_data` and attributes in sync.

Reply:
Oh, I see... Let's leave this as is for this PR then if you like, but I think it's worth some thought for the future. In general, storing the same information in two places is really difficult to maintain--it might actually make sense for Mindtpy to *not* follow GDPopt's design in terms of divorcing the `solve_data` object completely, then. I'm assuming that you pass the solve_data object to solver callbacks? Is there any way to pass the solver class instead, or no? Or maybe you could store a solve_data like object on the base class and all the algorithms could use that (including single-tree)? I'm not at all sure--these are just the first things that come to mind.

Reply:
We can discuss this for a future PR. The splitting of the solve data makes a lot of sense for multi-tree algorithms. Single-tree is a little different, and the data structures and their passing is not as clean as with the multi-tree. This might require some extra thought.